### ISSUE TITLE
inst.refresh() causing `Maximum call stack size exceeded`

### ISSUE THREAD

**Issue by romant** [acidb/mobiscroll#369]
using onMonthLoaded, and setting it to an array

```
onMonthLoaded: function (year, month, inst)
        {
        inst.settings.marked = events;
        inst.refresh();
        }
```

This is practically out of the examples provided; only am generating the array prior. In this case it's set to:

```
    var events = [  
            { d: '12/31', color: 'rgb(255,0,0)' }, 
            { d: '12/24', color: 'rgb(255,0,0)'}, 
            { d: '12/25', color: 'rgb(255,0,0)'},
            { d: new Date(), color: 'rgb(255,0,0)'},
            { d: new Date(), color: 'rgb(255,0,0)'},
        ];
```

The `inst.refresh()` presence causes a `Maximum call stack size exceeded` error to be thrown.

If after the function call, I instead use `$("#calendar").mobiscroll('refresh')` - it goe through without a hitch.

Why the difference in behaviour?

Many thanks.

PS:

no libraries are loaded beyond mobiscroll. Literally an empty, vanilla playground.

**Comment by dioslaska** [comment]
Since v2.17.0 the `refresh` method will also trigger the `onMonthLoaded` event (#336).
So, calling `refresh` method inside the `onMonthLoaded` event will result in an infinite loop.
As a workaround you can use a flag to prevent this:

    onMonthLoaded: function (year, month, inst) {
        if (!inst.refreshing) {
            inst.settings.marked = events;
            inst.refreshing = true;
            inst.refresh();
            inst.refreshing = false;
        }
    }

**Comment by dioslaska** [comment]
Fixed in v2.17.2

Use the [onMonthLoading](http://docs.mobiscroll.com/2-17-2/calendar#!event-onMonthLoading) event in combination with the [redraw](http://docs.mobiscroll.com/2-17-2/calendar#!method-redraw) method.

The main differences:

  - the `onMonthLoading` event is triggered before the markup is generated, so if marked days or events are not added async, refresh or redraw is not needed
  - the `redraw` method also regenerates the month markup, but data won't be reloaded and the `onMonthLoading` event will not be triggered

Sync:

    onMonthLoading: function (year, month, inst) {
        inst.settings.marked = events;
    }

Async:

    onMonthLoading: function (year, month, inst) {
        $.ajax({
            url: 'http://www.example.com/getMarkedDays?year=' + year + '&month=' + month,
            dataType: 'json',
            success : function (data) {

                inst.settings.marked = [];

                for (var i = 0; i < data.length; i++) {
                    inst.settings.marked.push({
                        d: new Date(data[i].date), // Make sure that a javascript date object is passed
                        text: data[i].text
                    });
                }

                // Redraw the view to display new data
                inst.redraw();
            }
        });
    }
